src/python/pants/backend/go/target_type_rules.py

```diff
@@ -24,8 +46,103 @@ async def inject_go_module_dependency(request: InjectGoModuleDependency) -> Inje
         return InjectedDependencies()
 
 
+class InferGoPackageDependenciesRequest(InferDependenciesRequest):
+    infer_from = GoPackageSources
+
+
+@rule
+async def infer_go_dependencies(
+    request: InferGoPackageDependenciesRequest,
+    goroot_imports: ResolvedImportPathsForGoLangDistribution,
+) -> InferredDependencies:
+    this_go_package = await Get(
+        ResolvedGoPackage, ResolveGoPackageRequest(request.sources_field.address)
+    )
+
+    # Obtain all go_package targets under this package's go_module.
+    assert this_go_package.module_address is not None
+    spec_path = this_go_package.module_address.spec_path
+    address_specs = [
+        MaybeEmptySiblingAddresses(spec_path),
+        MaybeEmptyDescendantAddresses(spec_path),
+    ]
+    candidate_targets = await Get(UnexpandedTargets, AddressSpecs(address_specs))
+    go_package_targets = [
+        tgt
+        for tgt in candidate_targets
+        if tgt.has_field(GoPackageSources)
+        and not tgt.address.is_file_target
+        and tgt.address != this_go_package.address
+    ]
+
+    # Find all go_external_modules in the repo and map their import paths to their address.
+    candidate_go_external_module_targets = await Get(
+        UnexpandedTargets, AddressSpecs([MaybeEmptyDescendantAddresses("")])
+    )
+    go_external_module_targets = [
+        tgt for tgt in candidate_go_external_module_targets if isinstance(tgt, GoExternalModule)
+    ]
+    third_party_import_path_to_address = {
+        tgt[GoImportPath].value: tgt.address for tgt in go_external_module_targets
+    }
+
+    # Resolve all of the packages found.
+    first_party_import_path_to_address = {}
+    first_party_go_packages = await MultiGet(
+        Get(ResolvedGoPackage, ResolveGoPackageRequest(tgt.address)) for tgt in go_package_targets
+    )
+    for first_party_go_package in first_party_go_packages:
+        # Skip packages that are not part of this package's module.
+        # TODO: This requires that all first-party code in the monorepo be part of the same go_module. Will need
+        # figure out how multiple modules in a monorepo can interact.
+        if first_party_go_package.module_address != this_go_package.module_address:
+            continue
+
+        first_party_import_path_to_address[
+            first_party_go_package.import_path
+        ] = first_party_go_package.address
```

Comment:
The computation of both of these mappings should probably be extracted into a separate `@rule` (or two) to avoid recomputing these dicts per package.

Reply:
Will add TODO.

Reply:
Agreed, FYI that's because the `@rule` will get memoized. This will get you much better perf.

Fine to be a followup.